### DISCUSSION TITLE
How to use ELSA?

### DISCUSSION THREAD

**Discussion by MarvinNorway** [elsa-workflows/elsa-core#4008] [Design & tech]
We have a really complex workflow, it can be considered similar to building a house.

As a first step, a person submits an application for a building permit. This building permit needs to be sent to the correct department/person, depending on the type of building and location. It could also be a group of people (e.g. city planning department). The approver shall receive in email that a new application is awaiting review, and there shall be periodic reminders if no action has been taken. Once approved, certain tasks shall be created in our system, e.g. "Pour concrete", "Construct walls". Some of them can be handled in parallel (windows can be installed while electricity is being installed), while others require other activities to be completed first (e.g. you can't add a roof if there are no walls). We also can't create all tasks right away, as they will depend on decisions along the way - maybe it hasn't been decided yet if solar panels should be added. We don't want to add a "Add solar panels" task and close it as "not applicable". Instead, at an appropriate time during the workflow, when the decision is made that solar panels shall be installed, only then the relevant tasks would be created. There are also some review/approval steps along the way - can't start working on the roof if the walls were not built according to the correct standard.
And most importantly, we have tasks that shall be triggered automatically - for example, once all tasks related to the construction of walls have been completed, new tasks for the installation of windows shall be created with a due date of 10 working days from the time the walls were completed.
During this whole workflow the overall project moves between different statuses (application submitted, application approved/rejected, final review, done etc.).

It would be great if we could visualize this workflow, or even let the client design it in a visual designer, instead of just writing code.

We would prefer if we didn't have to build "simple" functionality like sending emails ourselves, but a lot of our logic requires updating entities in our SQL database, or possibly even invoking external APIs. ELSA looks very promising, but I couldn't quite understand how to translate the samples to a real-world application. It seemed like if we use ELSA endpoints, then ELSA would basically replicate the whole project status inside the workflow...? Basically, ELSA would keep a copy of all projects and tasks...? We would prefer if we could keep using our own entities, so we can build our own task list etc., and persist it the way we see fit.

Do we have to use custom activities?
We would also be open to paying for a professional consultation.

**Comment by sfmskywalker** [comment]
Including the link to [SO](https://stackoverflow.com/questions/76157975/how-to-use-elsa-workflows-or-other-workflow-library-in-real-life-application) for completeness.
